### PR TITLE
[SPARK-9795] Dynamic allocation: avoid double counting when killing same executor twice

### PR DESCRIPTION
**Pull request by andrewor14**
This is based on @KaiXinXiaoLei's changes in #7716.

The issue is that when someone calls `sc.killExecutor("1")` on the same executor twice quickly, then the executor target will be adjusted downwards by 2 instead of 1 even though we're only actually killing one executor. In certain cases where we don't adjust the target back upwards quickly, we'll end up with jobs hanging.

This is a common danger because there are many places where this is called:
- `HeartbeatReceiver` kills an executor that has not been sending heartbeats
- `ExecutorAllocationManager` kills an executor that has been idle
- The user code might call this, which may interfere with the previous callers

While it's not clear whether this fixes SPARK-9745, fixing this potential race condition seems like a strict improvement. I've added a regression test to illustrate the issue.

### CODE REVIEW

**Comment by andrewor14** [comment]
@vanzin @srowen could you have a look?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 1]
```diff
@@ -422,16 +422,19 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       logWarning(s"Executor to kill $id does not exist!")
     }
 
+    // If an executor is already pending to be removed, do not kill it again (SPARK-9795)
+    val executorsToKill = knownExecutors.filter { id => !executorsPendingToRemove.contains(id) }
```
Comment (harishreedharan) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Nit: It looks like `unknownExecutors` is useful only for that one log statement. If we don't need that log line, we could reduce the number of set copies and traversals. How about something like this (this is less scala-like, but reduces the number of traversals and copies):

```
val knownExecutors = new HashSet[String]
executorsIds.foreach { id =>
  if (executorDataMap.contains(id)) {
    knownExecutors ++ id
  }
}
```

**Comment by harishreedharan** [comment]
LGTM

**Comment by vanzin** [comment]
LGTM

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: That looks fine, but in this patch I wanted to limit the scope of the changes so I'm going to leave this as is.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by andrewor14** [comment]
retest this please, just in case

**Comment by andrewor14** [comment]
retest this please. Pretty sure I didn't change any SQL...

**Comment by andrewor14** [comment]
Alright, the latest commit actually passed tests so I'm going to merge this into master 1.5.

**Pull request closed without merging**